In [1]:
import pandas as pd

# 같은 폴더(현재 작업 디렉토리)에 있다고 가정
naive_path = "eval_top5_truth_naive_top5_absolute.csv"
onto_path  = "eval_top5_truth_onto_top5_absolute.csv"

# CSV 로드
df_naive = pd.read_csv(naive_path)
df_onto  = pd.read_csv(onto_path)

# 출력
print("=== Naive RAG 평가 데이터 ===")
display(df_naive.head())

print("=== Ontology RAG 평가 데이터 ===")
display(df_onto.head())

print("Naive shape:", df_naive.shape)
print("Ontology shape:", df_onto.shape)



=== Naive RAG 평가 데이터 ===


query_id                                       query  rank  \
0         1  광주광역시의회 교육문화 분야에서 제기된 학교 예산 집행 문제를 정리해 달라.     1   
1         1  광주광역시의회 교육문화 분야에서 제기된 학교 예산 집행 문제를 정리해 달라.     2   
2         1  광주광역시의회 교육문화 분야에서 제기된 학교 예산 집행 문제를 정리해 달라.     3   
3         1  광주광역시의회 교육문화 분야에서 제기된 학교 예산 집행 문제를 정리해 달라.     4   
4         1  광주광역시의회 교육문화 분야에서 제기된 학교 예산 집행 문제를 정리해 달라.     5   

   retrieval_score                                            segment  \
0         0.630531  �구 제1선거구,더불어민주당)\n안평환입니다.\n박지영 부교육감님 비롯해서 전체적으...   
1         0.629791  1항 2025년도 광주광역시 일반 및 특별회계 세입세출 예산안, 의사일정 제2항 2...   
2         0.623461  �위원회 소관 의사일정 제2항부터 제7항까지 이상 여섯 건을 일괄 상정하겠습니다.\...   
3         0.619056  에 따라 계상된 특별회계 예비비 3990만 원 입니다.\n결산 승인과정은 「지방자치...   
4         0.618246  광역시교육청 기금운용변경계획안을 일괄 상정합니다.\n(의사봉 3타)\n제안설명 듣겠...   

   fact_score  person_score  party_score  council_score  topic_score  \
0         1.0           NaN          NaN            1.0          9.0   
1         1.0           NaN          NaN            1.0          6.0   
2         1.0           NaN          NaN            1.0          6.0   
3         1.0           NaN          NaN            1.0          8.0   
4         1.0           NaN          NaN            1.0          7.0   

   overall_score error_type  \
0            0.9       NONE   
1            0.6       NONE   
2            0.6       NONE   
3            0.8       NONE   
4            0.7       NONE   

                                             comment  overall_mean_top5  
0  세그먼트는 광주광역시의회에서 교육문화 분야의 학교 예산 집행 문제를 다루고 있어 사...               0.72  
1  세그먼트는 광주광역시의회 관련 예산안 심의 내용을 포함하고 있어 대상 의회와 일치하...               0.72  
2  세그먼트는 광주광역시의회 예산결산특별위원회의 2021년 회계연도 결산 심사 내용을 ...               0.72  
3  세그먼트는 광주광역시의회에서 교육문화 분야의 학교 예산 집행 문제와 관련된 질의응답...               0.72  
4  세그먼트는 광주광역시교육청의 2023년도 교육비특별회계 추가경정예산안에 관한 내용으...               0.72

=== Ontology RAG 평가 데이터 ===


query_id                                       query  rank  \
0         1  광주광역시의회 교육문화 분야에서 제기된 학교 예산 집행 문제를 정리해 달라.     1   
1         1  광주광역시의회 교육문화 분야에서 제기된 학교 예산 집행 문제를 정리해 달라.     2   
2         1  광주광역시의회 교육문화 분야에서 제기된 학교 예산 집행 문제를 정리해 달라.     3   
3         1  광주광역시의회 교육문화 분야에서 제기된 학교 예산 집행 문제를 정리해 달라.     4   
4         1  광주광역시의회 교육문화 분야에서 제기된 학교 예산 집행 문제를 정리해 달라.     5   

   retrieval_score                                            segment  \
0         0.535340  위원장 신수정\n예, 김요성 문화체육실장 수고하셨습니다.\n다음은 검토보고 듣겠습니...   
1         0.533879  위원장 박희율\n김환식 부교육감님, 수고하셨습니다.\n다음은 전문위원의 검토가 있습...   
2         0.532323  위원장 홍기월\n의석을 정돈하여 주시기 바랍니다. 회의를 속개하겠습니다.\n(의사봉...   
3         0.529819  (위원장 의사봉 없음)\n먼저 전문위원의 보고사항을 듣도록 하겠습니다. 전문위원 나...   
4         0.527313  위원장 신수정\n성원이 되었으므로 회의를 속개하겠습니다.\n(의사봉 3타)\n오전에...   

   fact_score  person_score  party_score  council_score  topic_score  \
0         1.0           NaN          NaN            1.0          8.0   
1         1.0           NaN          NaN            1.0          9.0   
2         1.0           NaN          NaN            1.0          9.0   
3         1.0           NaN          NaN            1.0          7.0   
4         1.0           NaN          NaN            1.0          8.0   

   overall_score error_type  \
0            0.8       NONE   
1            0.9       NONE   
2            0.9       NONE   
3            0.7       NONE   
4            0.8       NONE   

                                             comment  overall_mean_top5  
0  세그먼트는 광주광역시의회 문화체육실 소관 예산안 검토보고로, 광주광역시의회에 해당하...               0.82  
1  세그먼트는 광주광역시의회 관련 교육문화 분야의 학교 예산 집행 문제를 상세히 다루고...               0.82  
2  세그먼트는 광주광역시의회에서 교육문화 분야의 학교 예산 집행 문제에 대해 심도 있게...               0.82  
3  세그먼트는 광주광역시의회에서 교육 관련 예산안이 제출되고 심사되는 과정을 정확히 언...               0.82  
4  세그먼트 S는 광주광역시의회에서 교육문화 분야 관련 질의와 답변으로 구성되어 있으며...               0.82

Naive shape: (500, 14)
Ontology shape: (695, 14)


In [7]:
# ================================================
# 온톨로지/나이브 RAG 평가 데이터 전처리 셀
#  - 사실성 오류 정의: WRONG_COUNCIL / WRONG_PERSON / WRONG_PARTY
#  - 온톨로지 EMPTY_SEGMENT 질의는 아예 제외
#  - 해당 질의의 Naive 결과는 사실성 오류 + 관련성 0점 패널티
# ================================================

# 0) 기본 설정
FACT_ERROR_TYPES = {"WRONG_COUNCIL", "WRONG_PERSON", "WRONG_PARTY"}

# -------------------------------
# 1) WRONG_* 사실성 오류 개수 파악 (원본 기준)
# -------------------------------
naive_total_before = len(df_naive)
onto_total_before  = len(df_onto)

mask_naive_wrong = df_naive["error_type"].isin(FACT_ERROR_TYPES)
mask_onto_wrong  = df_onto["error_type"].isin(FACT_ERROR_TYPES)

num_naive_wrong = int(mask_naive_wrong.sum())
num_onto_wrong  = int(mask_onto_wrong.sum())

print("=== WRONG_* 사실성 오류 개수 (원본 기준) ===")
print(f"Naive  WRONG_* 단락 수: {num_naive_wrong} / {naive_total_before}")
print(f"Onto   WRONG_* 단락 수: {num_onto_wrong} / {onto_total_before}")
print("--------------------------------------------")

# -------------------------------
# 2) 온톨로지 EMPTY_SEGMENT 질의 파악 및 제거
# -------------------------------
mask_empty_onto = df_onto["error_type"] == "EMPTY_SEGMENT"
empty_qids = sorted(df_onto.loc[mask_empty_onto, "query_id"].unique())
empty_qid_set = set(empty_qids)
num_empty_rows_onto = int(mask_empty_onto.sum())

print(f"EMPTY_SEGMENT 질의 개수: {len(empty_qids)}")
print("EMPTY_SEGMENT query_id 목록:", empty_qids)
print(f"EMPTY_SEGMENT에 해당하는 Ontology 단락 수: {num_empty_rows_onto}")
print("--------------------------------------------")

# 온톨로지 데이터에서 EMPTY_SEGMENT 행 아예 제거
df_onto = df_onto.loc[~mask_empty_onto].copy()
onto_total_after = len(df_onto)

# -------------------------------
# 3) fact_ok 생성
#    - Ontology: WRONG_* 만 fact_ok = 0, 나머지는 1
#    - Naive   : WRONG_* 또는 EMPTY_SEGMENT 질의에 속하는 행 → fact_ok = 0
# -------------------------------

# (1) Ontology fact_ok
df_onto["fact_ok"] = df_onto["error_type"].apply(
    lambda e: 0 if e in FACT_ERROR_TYPES else 1
)

onto_fact_error = int((df_onto["fact_ok"] == 0).sum())

print("[Ontology] fact_ok 생성 후")
print(f"  사실성 오류(fact_ok=0) 단락 수: {onto_fact_error} / {onto_total_after}")
print("--------------------------------------------")

# (2) Naive fact_ok
mask_naive_empty = df_naive["query_id"].isin(empty_qid_set)

# 이전에 fact_ok가 있었다면 백업
if "fact_ok" in df_naive.columns:
    df_naive["fact_ok_before"] = df_naive["fact_ok"]
else:
    df_naive["fact_ok_before"] = None

# 기본값 1로 초기화 후, WRONG_* 또는 EMPTY_SEGMENT 질의는 0
df_naive["fact_ok"] = 1
df_naive.loc[mask_naive_wrong | mask_naive_empty, "fact_ok"] = 0

naive_fact_error = int((df_naive["fact_ok"] == 0).sum())

# 겹치는 경우(WRONG_* 이면서 EMPTY_SEGMENT 질의)도 확인
mask_naive_both       = mask_naive_wrong & mask_naive_empty
mask_naive_wrong_only = mask_naive_wrong & ~mask_naive_empty
mask_naive_empty_only = mask_naive_empty & ~mask_naive_wrong

num_naive_both       = int(mask_naive_both.sum())
num_naive_wrong_only = int(mask_naive_wrong_only.sum())
num_naive_empty_only = int(mask_naive_empty_only.sum())

print("[Naive] fact_ok 생성 후")
print(f"  사실성 오류(fact_ok=0) 단락 수: {naive_fact_error} / {naive_total_before}")
print("  ├ WRONG_* 만 해당 (EMPTY_SEGMENT 아님):", num_naive_wrong_only)
print("  ├ EMPTY_SEGMENT 질의만 해당          :", num_naive_empty_only)
print("  └ 둘 다 해당(WRONG_* ∧ EMPTY_SEGMENT):", num_naive_both)
print("    (위 세 개의 합이 fact_ok=0 총 개수이며, 각 행은 한 번만 카운트됨)")
print("--------------------------------------------")

# -------------------------------
# 4) Naive 관련성 점수 패널티 적용
#    - EMPTY_SEGMENT 질의의 모든 행: topic_score, overall_score = 0.0
#    - 변화 전 점수는 topic_before, overall_before에 백업
# -------------------------------
df_naive["topic_before"]   = df_naive["topic_score"]
df_naive["overall_before"] = df_naive["overall_score"]

df_naive.loc[mask_naive_empty, "topic_score"]   = 0.0
df_naive.loc[mask_naive_empty, "overall_score"] = 0.0

penalty_rows = int(mask_naive_empty.sum())

print(f"[Naive] EMPTY_SEGMENT 질의 패널티 적용된 단락 수: {penalty_rows}")
print("--------------------------------------------")

# -------------------------------
# 5) 최종 요약 출력
# -------------------------------
print("=== 최종 요약 ===")
print(f"Ontology 단락 수:  전처리 전 {onto_total_before}  → 전처리 후 {onto_total_after}")
print(f"Naive   단락 수:  전처리 전 {naive_total_before} → 전처리 후 {naive_total_before} (행 수는 동일, 점수/flag만 변경)")
print(f"Ontology 사실성 오류(fact_ok=0) 단락 수: {onto_fact_error}")
print(f"Naive   사실성 오류(fact_ok=0) 단락 수: {naive_fact_error}")
print("--------------------------------------------")

print("[Ontology] 사실성 오류 행 예시:")
display(df_onto.loc[df_onto["fact_ok"] == 0, ["query_id", "rank", "error_type", "fact_ok"]].head(10))

print("[Naive] 사실성 오류 행 예시:")
display(df_naive.loc[df_naive["fact_ok"] == 0, ["query_id", "rank", "error_type", "fact_ok"]].head(10))

print("[Naive] EMPTY_SEGMENT 질의 패널티 행 예시:")
display(
    df_naive.loc[mask_naive_empty, [
        "query_id", "rank",
        "topic_before", "topic_score",
        "overall_before", "overall_score",
        "error_type", "fact_ok"
    ]].head(10)
)

=== WRONG_* 사실성 오류 개수 (원본 기준) ===
Naive  WRONG_* 단락 수: 153 / 500
Onto   WRONG_* 단락 수: 43 / 695
--------------------------------------------
EMPTY_SEGMENT 질의 개수: 15
EMPTY_SEGMENT query_id 목록: [42, 44, 46, 48, 50, 52, 53, 54, 55, 57, 58, 59, 60, 67, 86]
EMPTY_SEGMENT에 해당하는 Ontology 단락 수: 85
--------------------------------------------
[Ontology] fact_ok 생성 후
  사실성 오류(fact_ok=0) 단락 수: 43 / 610
--------------------------------------------
[Naive] fact_ok 생성 후
  사실성 오류(fact_ok=0) 단락 수: 161 / 500
  ├ WRONG_* 만 해당 (EMPTY_SEGMENT 아님): 86
  ├ EMPTY_SEGMENT 질의만 해당          : 8
  └ 둘 다 해당(WRONG_* ∧ EMPTY_SEGMENT): 67
    (위 세 개의 합이 fact_ok=0 총 개수이며, 각 행은 한 번만 카운트됨)
--------------------------------------------
[Naive] EMPTY_SEGMENT 질의 패널티 적용된 단락 수: 75
--------------------------------------------
=== 최종 요약 ===
Ontology 단락 수:  전처리 전 695  → 전처리 후 610
Naive   단락 수:  전처리 전 500 → 전처리 후 500 (행 수는 동일, 점수/flag만 변경)
Ontology 사실성 오류(fact_ok=0) 단락 수: 43
Naive   사실성 오류(fact_ok=0) 단락 수: 161
--------------------

query_id  rank     error_type  fact_ok
34          7     5  WRONG_COUNCIL        0
36          8     2  WRONG_COUNCIL        0
53         11     4  WRONG_COUNCIL        0
54         11     5  WRONG_COUNCIL        0
55         12     1  WRONG_COUNCIL        0
64         13     5  WRONG_COUNCIL        0
75         16     1  WRONG_COUNCIL        0
88         18     4  WRONG_COUNCIL        0
104        21     5  WRONG_COUNCIL        0
109        22     5  WRONG_COUNCIL        0

[Naive] 사실성 오류 행 예시:


query_id  rank     error_type  fact_ok
7          2     3  WRONG_COUNCIL        0
9          2     5  WRONG_COUNCIL        0
12         3     3  WRONG_COUNCIL        0
14         3     5  WRONG_COUNCIL        0
20         5     1  WRONG_COUNCIL        0
22         5     3  WRONG_COUNCIL        0
23         5     4  WRONG_COUNCIL        0
28         6     4  WRONG_COUNCIL        0
33         7     4  WRONG_COUNCIL        0
39         8     5  WRONG_COUNCIL        0

[Naive] EMPTY_SEGMENT 질의 패널티 행 예시:


query_id  rank  topic_before  topic_score  overall_before  overall_score  \
205        42     1           0.0          0.0             0.0            0.0   
206        42     2           0.0          0.0             0.0            0.0   
207        42     3           0.0          0.0             0.0            0.0   
208        42     4           0.0          0.0             0.0            0.0   
209        42     5           0.0          0.0             0.0            0.0   
215        44     1           0.0          0.0             0.0            0.0   
216        44     2           0.0          0.0             0.0            0.0   
217        44     3           0.0          0.0             0.0            0.0   
218        44     4           0.0          0.0             0.0            0.0   
219        44     5           0.0          0.0             0.0            0.0   

       error_type  fact_ok  
205  WRONG_PERSON        0  
206  WRONG_PERSON        0  
207  WRONG_PERSON        0  
208  WRONG_PERSON        0  
209  WRONG_PERSON        0  
215  WRONG_PERSON        0  
216  WRONG_PERSON        0  
217  WRONG_PERSON        0  
218  WRONG_PERSON        0  
219  WRONG_PERSON        0


# 📘 **사실성 전처리 결과 해설 (Korean)**

## 1. 사실성 오류 기준 재정의

이번 분석에서는 기존 LLM 평가의 `fact_score`를 그대로 사용하지 않고,
**구조화된 사실 검증 오류(WRONG_COUNCIL, WRONG_PERSON, WRONG_PARTY)**만을
**실질적인 사실성 오류(fact_ok=0)**로 간주하였다.

이는 연구 목적상,

* **지자체가 올바른가?**
* **해당 의원이 맞는가?**
* **정당 정보가 맞는가?**

같은 핵심 사실 검증을 더 정확히 반영하기 위한 선택이다.

---

## 2. EMPTY_SEGMENT 질의의 처리

온톨로지 기반 RAG에서는 특정 15개 질의에서 **어떠한 세그먼트도 찾지 못함(EMPTY_SEGMENT)**을 확인했다.

이는 해당 질의에 대해 **의회 기록 내부에 일치하는 단락이 존재하지 않음**을 의미한다.

반면 Naive RAG는 같은 질의에 대해 **75개의 세그먼트를 “찾아왔다”고 판단했으나**,
이는 **존재하지 않는 정보를 세그먼트로 잘못 검색한 치명적인 사실성 오류**이다.

따라서 Naive RAG에서는 다음 처리가 적용되었다.

### ✔ EMPTY_SEGMENT 질의에 대해:

* 관련성 점수(topic_score, overall_score) = **0점 패널티**
* fact_ok = **0 (사실성 오류)**

온톨로지 RAG에서는 해당 질의 85개 행이 완전히 제거되어
**평가 분모에서 제외**되었다.

---

## 3. 전처리 후 사실성 오류 변화

### 🔹 WRONG_* 사실성 오류 개수(전처리 전)

| 모델           | WRONG_* 오류 / 전체 | 비율    |
| ------------ | --------------- | ----- |
| Naive RAG    | 153 / 500       | 30.6% |
| Ontology RAG | 43 / 695        | 6.2%  |

---

## 4. 전처리 후 단락 수 변화

| 모델           | 전처리 전 | 전처리 후        | 변화                        |
| ------------ | ----- | ------------ | ------------------------- |
| Ontology RAG | 695   | **610**      | EMPTY_SEGMENT 85개 제거      |
| Naive RAG    | 500   | **500 (동일)** | 단락 삭제 없음(점수와 fact_ok만 수정) |

---

## 5. 전처리 후 사실성 오류 개수

| 모델       | fact_ok=0 개수 | 세부 항목                                   |
| -------- | ------------ | --------------------------------------- |
| Ontology | **43**       | WRONG_* 오류만 포함                          |
| Naive    | **161**      | WRONG_* 86 + EMPTY_SEGMENT 8 + 중복 영역 67 |

※ Naive의 중복 67개는 WRONG_* 오류이면서 동시에 EMPTY_SEGMENT 질의를 포함한 경우이며,
한 번만 카운트되도록 정리됨.

---

## 6. 해석 요약

### ✔ Ontology RAG

* 전체 695개 중 85개는 단락 자체가 없으므로 제외됨
* 남은 610개 중 사실 오류는 43개
* **사실성 오류율: 43 / 610 ≈ 7.0%**

### ✔ Naive RAG

* 전체 500개 단락 중 사실 오류는 161개
* **사실성 오류율: 161 / 500 = 32.2%**

즉, **Naive RAG는 온톨로지 기반 RAG보다 약 4.6배 더 많은 사실성 오류**를 생성하였다.

---

# 📗 **Interpretation of Results (English)**

## 1. Redefinition of factual correctness

The original LLM-based `fact_score` was not used as-is.
Instead, factual correctness was redefined using a structured, rule-based criterion:

**Only the following are treated as factual errors (`fact_ok=0`):**

* WRONG_COUNCIL
* WRONG_PERSON
* WRONG_PARTY

This definition aligns better with the core factual checks in the legislative-minute domain.

---

## 2. Handling of EMPTY_SEGMENT queries

The ontology-based RAG completely failed to retrieve any segment for **15 queries**,
meaning **no relevant content exists in the original minutes**.

In contrast, Naive RAG retrieved **75 segments** for these invalid queries,
which constitutes **a critical factual error**—retrieving segments for topics that do not exist.

### ✔ Therefore:

* Ontology RAG: these 85 rows were **removed** before evaluation
* Naive RAG:

  * topical relevance set to **0**
  * fact_ok set to **0**
  * rows kept (for penalty accounting)

---

## 3. Factual error counts (before preprocessing)

| Model        | WRONG_* errors | Total rows | Error rate |
| ------------ | -------------- | ---------- | ---------- |
| Naive RAG    | 153            | 500        | 30.6%      |
| Ontology RAG | 43             | 695        | 6.2%       |

---

## 4. Number of rows after preprocessing

| Model    | Before | After   | Change                        |
| -------- | ------ | ------- | ----------------------------- |
| Ontology | 695    | **610** | removed 85 EMPTY_SEGMENT rows |
| Naive    | 500    | **500** | scores/flags updated only     |

---

## 5. Final factual error counts (after preprocessing)

| Model    | fact_ok=0 | Breakdown                                      |
| -------- | --------- | ---------------------------------------------- |
| Ontology | **43**    | WRONG_* only                                   |
| Naive    | **161**   | WRONG_* 86 + EMPTY_SEGMENT 8 + intersection 67 |

(The intersection rows are counted once.)

---

## 6. Interpretation

* Ontology RAG factual error rate: **43 / 610 ≈ 7%**
* Naive RAG factual error rate: **161 / 500 = 32.2%**

Thus, **Naive RAG yields approximately 4.6× more factual errors** than the ontology-based RAG.

In [10]:
# Naive / Ontology 사실성 오류율 비교 테이블 만들기

# 1) 기본 집계
n_naive   = len(df_naive)
n_onto    = len(df_onto)

err_naive = (df_naive["fact_ok"] == 0).sum()
err_onto  = (df_onto["fact_ok"] == 0).sum()

ok_naive  = n_naive - err_naive
ok_onto   = n_onto - err_onto

rate_naive = err_naive / n_naive if n_naive > 0 else float("nan")
rate_onto  = err_onto  / n_onto  if n_onto  > 0 else float("nan")

# 2) 비교용 DataFrame 생성
summary = pd.DataFrame({
    "model":        ["Naive RAG", "Ontology RAG"],
    "n_segments":   [n_naive,      n_onto],
    "n_errors":     [err_naive,    err_onto],
    "n_correct":    [ok_naive,     ok_onto],
    "error_rate":   [rate_naive,   rate_onto],
    "error_rate_%": [rate_naive * 100, rate_onto * 100],
})

print("=== Fact-level error summary ===")
display(summary)


=== Fact-level error summary ===


model  n_segments  n_errors  n_correct  error_rate  error_rate_%
0     Naive RAG         500       161        339    0.322000      32.20000
1  Ontology RAG         610        43        567    0.070492       7.04918

# 📘 **Fact-level Error Summary — Interpretation**

아래 표는 사실성 기준(fact_ok)에 따라 **각 RAG 시스템이 얼마나 정확한 단락을 가져왔는지**를 비교한 것이다.

| 모델           | 전체 단락 수 | 사실 오류 단락 수 | 정확 단락 수 | 오류율        |
| ------------ | ------- | ---------- | ------- | ---------- |
| Naive RAG    | 500     | 161        | 339     | **32.20%** |
| Ontology RAG | 610     | 43         | 567     | **7.05%**  |

### 🔍 주요 해석

1. **Ontology RAG는 사실 오류율이 32.20% → 7.05%로 크게 감소했다.**
   이는 약 **-25.15%p**, 비율로는 **~78% 감소**에 해당하는 매우 큰 개선이다.

2. **정확한 단락 수는 339 → 567로 증가했다.**
   Ontology RAG는 더 많은 단락을 유지하면서도, 오류 발생을 큰 폭으로 줄였다.

3. **Naive RAG에서 발생한 사실성 문제(WRONG_COUNCIL / WRONG_PERSON / WRONG_PARTY) 대부분이 Ontology 필터링 단계에서 제거되었다.**

4. **Ontology RAG의 core 효과**

   * 인물/정당/의회 매칭 오류를 구조적으로 제거
   * 문제 쿼리에 대해 불필요한 문맥 확장을 방지
   * Embedding-기반 의미 검색의 “망각 오류(noise)”를 보완

### 📌 결론

> **사실성(factual correctness) 측면에서 Ontology 기반 RAG는 Naive RAG보다 훨씬 안정적이며, 구조적 필터링을 통해 오류 발생 가능성을 근본적으로 억제한다.**

---

The following table compares **how accurately each RAG system retrieved segments**, based on whether each segment was classified as factually correct (fact_ok = 1) or incorrect (fact_ok = 0).

| Model        | # Segments | # Fact Errors | # Correct | Error Rate |
| ------------ | ---------- | ------------- | --------- | ---------- |
| Naive RAG    | 500        | 161           | 339       | **32.20%** |
| Ontology RAG | 610        | 43            | 567       | **7.05%**  |

### 🔍 Key Interpretation

1. **Ontology RAG dramatically reduces the factual error rate**
   from **32.20% down to 7.05%**,
   a reduction of **25.15 percentage points**, or roughly **−78% relative error reduction**.

2. **The number of correct segments increases from 339 to 567**,
   meaning Ontology RAG retrieves **more segments** while **making far fewer factual mistakes**.

3. **Most fact errors that occurred in the Naive RAG**
   (WRONG_COUNCIL / WRONG_PERSON / WRONG_PARTY)
   are systematically eliminated by the ontology-based filtering layer.

4. **Core benefit of Ontology RAG**

   * Prevents mismatching of person/party/council entities
   * Reduces semantic noise inherent in embedding-based retrieval
   * Ensures query–segment alignment at the structural level

### 📌 Conclusion

> **In terms of factual consistency, Ontology-based RAG is far more robust and reliable than Naive RAG, thanks to structural filtering that prevents critical retrieval errors.**

In [14]:
# 기존 range/describe 출력
print("Naive topic_score range:", df_naive['topic_score'].min(), df_naive['topic_score'].max())
print("Ontology topic_score range:", df_onto['topic_score'].min(), df_onto['topic_score'].max())

print("\nNaive topic_score describe:")
display(df_naive['topic_score'].describe())

print("\nOntology topic_score describe:")
display(df_onto['topic_score'].describe())

# ------------------------------------------
# 추가: 평균 비교 표 생성
# ------------------------------------------

max_score = 10  # topic_score의 만점

comparison_df = pd.DataFrame({
    "Model": ["Naive RAG", "Ontology RAG"],
    "Mean Topic Score": [
        df_naive["topic_score"].mean(),
        df_onto["topic_score"].mean()
    ],
    "Max Score": [max_score, max_score]
})

print("\n=== Topic Score Mean Comparison ===")
display(comparison_df)

Naive topic_score range: 0.0 10.0
Ontology topic_score range: 2.0 10.0

Naive topic_score describe:


count    500.000000
mean       5.772000
std        2.948458
min        0.000000
25%        4.000000
50%        7.000000
75%        8.000000
max       10.000000
Name: topic_score, dtype: float64


Ontology topic_score describe:


count    610.000000
mean       6.537705
std        2.030002
min        2.000000
25%        5.000000
50%        7.000000
75%        8.000000
max       10.000000
Name: topic_score, dtype: float64


=== Topic Score Mean Comparison ===


Model  Mean Topic Score  Max Score
0     Naive RAG          5.772000         10
1  Ontology RAG          6.537705         10

# 📘 **Topic Score Mean Comparison — Interpretation**


다음은 Naive RAG와 Ontology RAG의 **topic_score(관련성 점수)** 평균을 비교한 결과이다.

| 모델           | 평균 관련성 점수     | 만점 |
| ------------ | ------------- | -- |
| Naive RAG    | **5.77 / 10** | 10 |
| Ontology RAG | **6.54 / 10** | 10 |

### 🔍 주요 해석

1. **Ontology RAG의 평균 관련성 점수는 5.77 → 6.54로 상승했다.**
   이는 단순 평균 기준 **약 0.77점 증가**, 비율로는 **약 +13.3% 상승**이다.

2. **Naive RAG는 0점부터 10점까지 넓은 분포를 보였지만**,
   Ontology RAG는 **최소 점수가 2점**으로 훨씬 안정적인 점수 분포를 보인다.
   → 즉, Ontology RAG는 “큰 실패(0점대 검색)”가 거의 없다.

3. Ontology 기반 필터링은

   * 쿼리와 무관한 문맥(irrelevant)을 사전에 차단하고
   * 검색 공간을 질적으로 압축하며
   * 의미 기반 검색의 노이즈를 줄여
     보다 일관된 고관련성(topical relevance) 결과를 제공한다.

### 📌 결론

> **Ontology RAG는 관련성 측면에서 더 안정적이며 평균 점수도 높다.
> Embedding 기반의 Naive RAG에서 발생하는 비관련(non-topic) 검색 문제를 구조적으로 해결한 결과이다.**

---
Below is the comparison of the **average topic_score (topical relevance)** for the two RAG systems.

| Model        | Mean Topic Score | Max Score |
| ------------ | ---------------- | --------- |
| Naive RAG    | **5.77 / 10**    | 10        |
| Ontology RAG | **6.54 / 10**    | 10        |

### 🔍 Key Interpretation

1. **Ontology RAG improves the mean topic score**
   from **5.77 to 6.54**,
   a gain of **+0.77 points**, or approximately **+13.3% improvement**.

2. **Naive RAG scores range from 0 to 10**,
   showing a wide spread that includes many low-relevance matches.
   In contrast, **Ontology RAG’s minimum score is 2**, indicating far fewer catastrophic failures.

3. The ontology filtering layer

   * blocks irrelevant or low-quality segments
   * reduces semantic noise
   * enforces structural consistency (person/party/council alignment)
     leading to more reliable and consistent topical retrieval.

### 📌 Conclusion

> **Ontology RAG delivers both higher and more stable topical relevance.
> Its structural filtering compensates for the weaknesses of purely embedding-based retrieval.**

In [20]:
# 오류율 (fact_ok = 0 비율)
naive_error_rate = (df_naive["fact_ok"] == 0).mean()
onto_error_rate  = (df_onto["fact_ok"] == 0).mean()

# topic_score 평균
naive_topic_mean = df_naive["topic_score"].mean()
onto_topic_mean  = df_onto["topic_score"].mean()

# 만점
MAX_SCORE = 10

# ----------------------------------
# 2) % 변동 계산 (만점 대비)
# ----------------------------------

naive_topic_pct = naive_topic_mean / MAX_SCORE * 100
onto_topic_pct  = onto_topic_mean / MAX_SCORE * 100
topic_pct_change = onto_topic_pct - naive_topic_pct  # %

error_rate_change = (onto_error_rate - naive_error_rate) * 100  # %p

# ----------------------------------
# 3) 표 구성 (문자열 포맷으로 보기 좋게)
# ----------------------------------

summary_df = pd.DataFrame({
    "Metric": ["Fact Error Rate", "Topic Score (avg)"],
    "Naive RAG": [
        f"{naive_error_rate*100:.2f}%",
        f"{naive_topic_mean:.2f}/{MAX_SCORE}"
    ],
    "Ontology RAG": [
        f"{onto_error_rate*100:.2f}%",
        f"{onto_topic_mean:.2f}/{MAX_SCORE}"
    ],
    "Change": [
        f"{error_rate_change:.2f}%",
        f"{topic_pct_change:.2f}%"
    ]
})

print("=== RAG 성능 비교 요약표 ===")
display(summary_df)

=== RAG 성능 비교 요약표 ===


Metric Naive RAG Ontology RAG   Change
0    Fact Error Rate    32.20%        7.05%  -25.15%
1  Topic Score (avg)   5.77/10      6.54/10    7.66%